Kernel/ACPI/MultiProcessorParser.cpp

```diff
@@ -0,0 +1,268 @@
+    auto region = MM.allocate_kernel_region(PhysicalAddress(0), PAGE_SIZE, "MultiProcessor Parser Floating Pointer Structure Finding", Region::Access::Read);
+    u16 ebda_seg = (u16) * ((uint16_t*)((region->vaddr().get() & PAGE_MASK) + 0x40e));
+    kprintf("MultiProcessor: Probing EBDA, Segment 0x%x\n", ebda_seg);
+
+    mp_floating_pointer = search_floating_pointer_in_ebda(ebda_seg);
+    if (mp_floating_pointer != nullptr)
+        return mp_floating_pointer;
+    return search_floating_pointer_in_bios_area();
+}
+
+MultiProcessor::FloatingPointer* MultiProcessorParser::search_floating_pointer_in_ebda(u16 ebda_segment)
+{
+    auto floating_pointer_region = MM.allocate_kernel_region(PhysicalAddress(page_base_of((u32)(ebda_segment << 4))), PAGE_ROUND_UP(1024), "MultiProcessor Parser floating_pointer Finding #1", Region::Access::Read, false, true);
+    char* p_floating_pointer_str = (char*)(PhysicalAddress(ebda_segment << 4).as_ptr());
+    for (char* floating_pointer_str = (char*)floating_pointer_region->vaddr().offset(offset_in_page((u32)(ebda_segment << 4))).as_ptr(); floating_pointer_str < (char*)(floating_pointer_region->vaddr().offset(offset_in_page((u32)(ebda_segment << 4))).get() + 1024); floating_pointer_str += 16) {
+#ifdef MUTLIPROCESSOR_DEBUG
+        dbg() << "MultiProcessor: Looking for floating pointer structure in EBDA @ V0x " << String::format("%x", floating_pointer_str) << ", P0x" << String::format("%x", p_floating_pointer_str);
+#endif
+        if (!strncmp("_MP_", floating_pointer_str, strlen("_MP_")))
+            return (MultiProcessor::FloatingPointer*)p_floating_pointer_str;
+        p_floating_pointer_str += 16;
+    }
+    return nullptr;
+}
+MultiProcessor::FloatingPointer* MultiProcessorParser::search_floating_pointer_in_bios_area()
+{
+    auto floating_pointer_region = MM.allocate_kernel_region(PhysicalAddress(page_base_of((u32)0xE0000)), PAGE_ROUND_UP(0xFFFFF - 0xE0000), "MultiProcessor Parser floating_pointer Finding #2", Region::Access::Read, false, true);
+    char* p_floating_pointer_str = (char*)(PhysicalAddress(0xE0000).as_ptr());
+    for (char* floating_pointer_str = (char*)floating_pointer_region->vaddr().offset(offset_in_page((u32)(0xE0000))).as_ptr(); floating_pointer_str < (char*)(floating_pointer_region->vaddr().offset(offset_in_page((u32)(0xE0000))).get() + (0xFFFFF - 0xE0000)); floating_pointer_str += 16) {
+#ifdef MUTLIPROCESSOR_DEBUG
+        dbg() << "MultiProcessor: Looking for floating pointer structure in BIOS area @ V0x " << String::format("%x", floating_pointer_str) << ", P0x" << String::format("%x", p_floating_pointer_str);
+#endif
+        if (!strncmp("_MP_", floating_pointer_str, strlen("_MP_")))
+            return (MultiProcessor::FloatingPointer*)p_floating_pointer_str;
+        p_floating_pointer_str += 16;
+    }
+    return nullptr;
+}
+
+Vector<unsigned>& MultiProcessorParser::get_pci_bus_ids()
+{
+    Vector<unsigned>& pci_bus_ids = *new Vector<unsigned>();
```

Comment:
Please don't abuse `operator new` and references like this.
References and raw pointers should only refer to objects kept alive by *someone else*.
Also, as there are no corresponding `delete` statements, objects allocated this way will leak.

Instead, simply use value semantics by returning a `Vector<unsigned>` and creating a `Vector<unsigned> on the stack:

```cpp
Vector<unsigned> MultiProcessorParser::get_pci_bus_ids()
{
    Vector<unsigned> pci_bus_ids;
    ....
    return pci_bus_ids;
}
```